File: node/shared/Service.ts
import express, { Application, Request, Response, NextFunction, Router } from 'express';
import http from 'http';
import https from 'https';
import fs from 'file-system';
import cors from 'cors';
import path from 'path';

import type { Service as Microservice } from '../shared/types/service.d.ts';
import redisClient from './redis/client';

export default class Service {

  public metadata: Microservice.Metadata;

  // express server
  private app: Application;
  private port: number;
  private router: Router;
  private server: http.Server | https.Server;

  // kafka
  private producer: any; 
  private consumer: any;
  
  constructor(config: Microservice.Config, metadata: Microservice.Metadata) {

    const { port, router, kafka } = config;
    
    this.app = express();
    this.port = port;
    this.router = router;
    this.metadata = metadata;

    this.configureMiddlewares();
    this.configureRoutes();
    this.configureKafka(kafka);
    this.configureRedis();
    this.createServer(config);
  }

  private configureMiddlewares(): void {
    this.app.use(express.json());
    this.app.use(express.urlencoded({ extended: true }));
    this.app.use(function (req: Request, res: Response, next: NextFunction) {
      res.setHeader('Access-Control-Allow-Origin', "*");
      res.setHeader('Access-Control-Allow-Methods', 'GET, POST, OPTIONS, PUT, PATCH, DELETE');
      res.setHeader('Access-Control-Allow-Headers', 'X-Requested-With,content-type');
      res.setHeader('Access-Control-Allow-Credentials', 'true');
      next();
    });
    this.app.use(cors());
    this.app.set('trust proxy', 1);
    this.app.disable('x-powered-by');
    this.app.use(require('cookie-parser')());
    this.app.use(require('body-parser').urlencoded({ extended: false }));
  }

  private configureRoutes(): void {
    const router = this.router; 
    this.app.use(router);
  }

  private configureKafka(kafka: any) {

    const { producer, consumer } = kafka;

    let options = kafka;
    delete options.producer; delete options.consumer;

    this.producer = new producer(options);
    this.consumer = new consumer(options);
  }

  private async configureRedis() {
    try {
      // Start Redis Client
      await redisClient.connect(); 
      return;
    } catch (error) {
      console.error(error);
    }
  }

  public createServer(options: Microservice.Config): void {
    const { ssl } = options;
    if (ssl && process.env.NODE_ENV !== "production" && process.env.DOCKER !== 'true') {
      const httpsOptions: https.ServerOptions = {
        key: fs.readFileSync(path.resolve(__dirname, ssl.key)),
        cert: fs.readFileSync(path.resolve(__dirname, ssl.cert))
      }
      this.server = https.createServer(httpsOptions, this.app);
    } else {
      this.server = http.createServer(this.app);
    }
  }
  
  public async start(): Promise<void> {
    try {

      // start express server
      this.server.listen(this.port, () => {
        console.log(`${this.metadata.name} service listening on port ${this.port}`);
      });

      // start kafka
      await this.producer.start();
      await this.consumer.start();

    } catch (error) {
      console.log(error);
    }
  }

  public async stop(): Promise<void> {
    try {
      // start express server
      this.server.close(() => {
        console.log(`${this.metadata.name} service closed on port ${this.port}`);
      });

      // start kafka
      await this.producer.shutdown();
      await this.consumer.shutdown();

    } catch (error) {
      console.log(error);
    }
  }

}

File: go/shared/Service.go
package service

import (
	"crypto/tls"
	"fmt"
	"net/http"
	"os"

	"github.com/NCSU-Microservice-Benchmarking/vision_middleware/go/shared/kafka"
	"github.com/NCSU-Microservice-Benchmarking/vision_middleware/go/shared/types"
	"github.com/gorilla/mux"
	"github.com/rs/cors"
)

type Service struct {
	Metadata types.ServiceMetadata
	App      *mux.Router
	Port     int
	Server   *http.Server
	Producer *kafka.Producer
	Consumer *kafka.Consumer
}

func NewService(config types.ServiceConfig, metadata types.ServiceMetadata) *Service {
	s := &Service{
		Metadata: metadata,
		Port:     config.Port,
	}

	s.App = mux.NewRouter()
	s.configureMiddlewares()
	s.configureRoutes()
	s.configureKafka(config.KafkaOptions)
	s.configureRedis()
	s.createServer(config)

	return s
}

func (s *Service) configureMiddlewares() {
	s.App.Use(cors.New(cors.Options{
		AllowedOrigins:   []string{"*"},
		AllowedMethods:   []string{"GET", "POST", "OPTIONS", "PUT", "PATCH", "DELETE"},
		AllowedHeaders:   []string{"X-Requested-With", "content-type"},
		AllowCredentials: true,
	}).Handler)
	s.App.Use(mux.CORSMethodMiddleware(s.App))
	s.App.Use(func(next http.Handler) http.Handler {
		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			fmt.Println("Logging middleware:", r.RequestURI)
			next.ServeHTTP(w, r)
		})
	})
}

func (s *Service) configureRoutes() {
	// Define your routes here using s.App
}

func (s *Service) configureKafka(kafkaOptions types.KafkaOptions) error {
	var err error // Define err variable here

	if kafkaOptions.Topics.Producer != "" {
		s.Producer, err = kafka.NewProducer(s.Metadata.Name, kafkaOptions)
		if err != nil {
			return err // Return error if producer creation fails
		}
	}

	// Then, define the producerCallback function using the same signature as ProducerCallback
	producerCallback := func(request types.Request, topic string) error {
		// Call the Producer's Send method
		return s.Producer.Send(request)
	}

	// Create the consumer with the correct callback
	s.Consumer, err = kafka.NewConsumer(s.Metadata.Name, kafkaOptions, producerCallback)
	if err != nil {
		return err // Return error if consumer creation fails
	}

	// No errors occurred, return nil
	return nil
}

func (s *Service) configureRedis() {
	// Implement Redis configuration if needed
}

func (s *Service) createServer(config types.ServiceConfig) {
	if config.SSL != nil && os.Getenv("NODE_ENV") != "production" && os.Getenv("DOCKER") != "true" {
		httpsOptions := &tls.Config{
			MinVersion: tls.VersionTLS12,
			CurvePreferences: []tls.CurveID{
				tls.CurveP521,
				tls.CurveP384,
				tls.CurveP256,
			},
		}

		s.Server = &http.Server{
			Addr:      fmt.Sprintf(":%d", s.Port),
			Handler:   s.App,
			TLSConfig: httpsOptions,
		}
	} else {
		s.Server = &http.Server{
			Addr:    fmt.Sprintf(":%d", s.Port),
			Handler: s.App,
		}
	}
}

func (s *Service) Start() error {
	fmt.Printf("%s service listening on port %d\n", s.Metadata.Name, s.Port)

	if s.Producer != nil {
		if err := s.Producer.Start(); err != nil {
			return err
		}
	}

	if err := s.Consumer.Start(); err != nil {
		return err
	}

	return s.Server.ListenAndServe()
}

func (s *Service) Stop() error {
	if err := s.Server.Close(); err != nil {
		return err
	}

	if s.Producer != nil {
		if err := s.Producer.Shutdown(); err != nil {
			return err
		}
	}

	if err := s.Consumer.Shutdown(); err != nil {
		return err
	}

	// Stop Redis here if needed

	return nil
}
